server/tests/Icepeak/Server/MultiSubscriptionSpec.hs

```diff
@@ -0,0 +1,340 @@
+withIcepeak = do
+  let
+    storageFile = "/tmp/icepeak.json"
+    config = defaultConfig { Icepeak.configDataFile = Just storageFile }
+  writeFile storageFile "{}"
+  logger <- Icepeak.newLogger config
+  core <- Icepeak.newCore config logger Nothing >>=
+    (\mbCore -> case mbCore of
+        Left err -> do expectationFailure ("Failed to create Core: " <> err)
+                       undefined
+        Right core -> pure core)
+  
+  let wsServer = Icepeak.acceptConnection core
+  application <- IcepeakHttp.new core
+
+  -- Thread that accepts websocket connections
+  webserverThread <- Async.async $ Icepeak.runServer logger wsServer application icepeakPort
+
+  -- Thread that processes the Icepeak.Command queue
+  commandLoopThread <- Async.async $ Icepeak.runCommandLoop core
+
+  -- Thread that posts updates to websocket clients
+  webSocketThread <- Async.async $ Icepeak.processUpdates core
+
+  pure $ Icepeak
+    { icepeakCore = core
+    , icepeakShutdown = do
+        mapM_ Async.cancel [webserverThread, commandLoopThread, webSocketThread ]
+        Directory.removeFile storageFile
+    }
+
+createDataSet :: Icepeak -> IO ()
+createDataSet icepeak = do
+  mapM_ (makeModification (icepeakCore icepeak))
+    [ Icepeak.Put ["A", "A"] "A"
+    , Icepeak.Put ["A", "B"] "B"
+    , Icepeak.Put ["B"] "B"
+    ]
+
+makeModification :: Icepeak.Core -> Icepeak.Modification -> IO ()
+makeModification core modification = do
+  waitMVar <- MVar.newEmptyMVar
+  Icepeak.enqueueCommand (Icepeak.Modify modification (Just waitMVar)) core
+  MVar.takeMVar waitMVar
+
+data CloseExpectation
+  = Unexpected String
+  | CloseCode Word16
+  deriving (Show, Eq)
+
+openReusableIcepeakConn :: (WS.Connection -> IO a) -> IO a
+openReusableIcepeakConn = WS.runClient "localhost" icepeakPort "/?method=reusable"
+
+sendJson :: WS.Connection -> Aeson.Value -> IO ()
+sendJson conn value = WS.sendDataMessage conn (WS.Text (Aeson.encode value) Nothing)
+
+expectDataMessage :: WS.Connection -> IO WS.DataMessage
+expectDataMessage conn = do
+  Right msg <- Async.race
+    (do Concurrent.threadDelay 100_000 -- 0.1 s
+        expectationFailure "Client expected a message, but did not receive anything for the duration of the 0.1s timeout")
+    (WS.receiveDataMessage conn)
+  pure msg
+
+expectNoMessage :: WS.Connection -> IO ()
+expectNoMessage conn = do
+  Left () <- Async.race
+    (Concurrent.threadDelay 10_000) -- 0.01s
+    (WS.receiveDataMessage conn >>=
+      (\p -> expectationFailure
+        $ "Client received an unexpected payload: " <> show p))
+  pure ()
+
+withResponseJson :: WS.Connection -> (Aeson.Value -> Expectation) -> Expectation
+withResponseJson conn jsonCheck = do
+  (WS.Text payload _) <- expectDataMessage conn
+  (Just json) <- pure $ Aeson.decode payload
+  jsonCheck json
+
+invalidPayloadsSpec :: SpecWith a
+invalidPayloadsSpec = describe "Opening and sending invalid payloads" $ do
+  it "should close connection upon invalid payload" $ const $ do
```

Comment:
This might be better suited as a
[`context`](https://hackage.haskell.org/package/hspec-2.11.7/docs/Test-Hspec.html#v:context)
so the individual `shouldBe` can be their own tests via `it`.